In [2]:
##############

# RUN ONLY ONCE

##############

### Downloading Data

In [ ]:
# Instead of running the cell below, you can directly go to the below link to download the data:

# https://www.kaggle.com/datasets/saicakella/commonwealth-of-australia-bureau-of-meteorology

# PLEASE ENSURE TO STORE THE FILES UNDERS datasets/original

# If you opt to run the cell below you need a kaggle account, then download the you kaggle api and store into the .kaggle folder. Please refer to instructions on the web for this.

In [5]:
# downloading data from kaggle private repo
!kaggle datasets download -d saicakella/commonwealth-of-australia-bureau-of-meteorology --unzip -p ../datasets/original

Dataset URL: https://www.kaggle.com/datasets/saicakella/commonwealth-of-australia-bureau-of-meteorology
License(s): unknown
  0%|                                               | 0.00/43.6M [00:00<?, ?B/s]
100%|██████████████████████████████████████| 43.6M/43.6M [00:00<00:00, 1.60GB/s]


### Importing required libraries

In [6]:
import numpy as np
import pandas as pd
import os

### Helper Functions

In [7]:
def list_dir(path:str) -> None:
    for directory, folders, files in os.walk(path):
        print(f'{directory} contain the following folders: {folders if folders else "NO FOLDERS FOUND IN DIRECTORY"}')
        print(f'{directory} contain the following files: {files if files else "NO FILES FOUND IN DIRECTORY"}')

### Loading Datasets + Exploration of Datasets + Clean-up/Creation of main dataset

In [8]:
list_dir("../datasets")

../datasets contain the following folders: ['final_datasets', 'original']
../datasets contain the following files: ['schema.sql']
../datasets/final_datasets contain the following folders: NO FOLDERS FOUND IN DIRECTORY
../datasets/final_datasets contain the following files: NO FILES FOUND IN DIRECTORY
../datasets/original contain the following folders: NO FOLDERS FOUND IN DIRECTORY
../datasets/original contain the following files: ['NT.csv', 'WA..csv', 'VIC.csv', 'AAT.csv', 'quality-metatdata.txt', 'Location.csv', 'description.csv', 'NSW.csv', 'SA.csv', 'TAS.csv', 'AET.csv', 'QLD.csv']


- state specific BOM data: 
    - 'NT.csv'
    - 'WA..csv'
    - 'VIC.csv'
    - 'AAT.csv'
    - 'NSW.csv'
    - 'SA.csv'
    - 'TAS.csv'
    - 'AET.csv'
    - 'QLD.csv'
- 'quality-metatdata.txt'
- 'Location.csv'
- 'description.csv'

#### quality-metadata

In [9]:
with open('../datasets/original/quality-metatdata.txt', 'r') as fp:
    print(fp.read())

* QUALITY FLAG DESCRIPTIONS
___________________________
 
Y: quality controlled and acceptable
N: not quality controlled
W: quality controlled and considered wrong
S: quality controlled and considered suspect
I: quality controlled and inconsistent with other known information
X: no quality information available



#### Location

In [10]:
location_df = pd.read_csv("../datasets/original/Location.csv")
location_df

Site                  Name      Lat      Long   State  \
0      1006          WYNDHAM AERO -15.5100  128.1503    W.A.   
1      1007      TROUGHTON ISLAND -13.7542  126.1485    W.A.   
2      3003        BROOME AIRPORT -17.9475  122.2352   W.A.    
3      3032            DERBY AERO -17.3706  123.6611    W.A.   
4      4019               MANDORA -19.7419  120.8433    W.A.   
..      ...                   ...      ...       ...     ...   
136  200283         WILLIS ISLAND -16.2878  149.9652  A.E.T.   
137  200284  COCOS ISLAND AIRPORT -12.1892   96.8344  A.E.T.   
138  200288   NORFOLK ISLAND AERO -29.0389  167.9408  A.E.T.   
139  300000                 DAVIS -68.5744   77.9672  A.A.T.   
140  300004      MACQUARIE ISLAND -54.4994  158.9369  A.A.T.   

                     Region  
0    Wyndham East Kimberley  
1                 Kimberley  
2                    Broome  
3      Derby West Kimberley  
4                    Broome  
..                      ...  
136                    N.A.  
137                    N.A.  
138                    N.A.  
139                    N.A.  
140                    N.A.  

[141 rows x 6 columns]

In [11]:
location_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 141 entries, 0 to 140
Data columns (total 6 columns):
 #   Column  Non-Null Count  Dtype  
---  ------  --------------  -----  
 0   Site    141 non-null    int64  
 1   Name    141 non-null    object 
 2   Lat     141 non-null    float64
 3   Long    141 non-null    float64
 4   State   141 non-null    object 
 5   Region  141 non-null    object 
dtypes: float64(2), int64(1), object(3)
memory usage: 6.7+ KB


In [12]:
location_df.columns

Index(['Site', 'Name', 'Lat', 'Long', 'State', 'Region'], dtype='object')

In [13]:
location_df = location_df.rename(columns=str.lower) # chaning column name to lower case

In [14]:
location_df['name'].nunique()

141

In [15]:
location_df.groupby(by=['lat','long']).ngroups

141

In [16]:
location_df['state'].unique()

array(['W.A.', 'W.A. ', 'N.T.', 'N.T. ', 'S.A.', 'QLD', 'N.S.W.', 'VIC',
       'TAS', 'A.E.T.', 'A.A.T.'], dtype=object)

In [17]:
np.sort(location_df['region'].unique())

array(['Aboriginal Shire of Kowanyama', 'Adelaide', 'Alice Springs',
       'Augusta - Margaret River', 'Ballarat', 'Banana', 'Barkly',
       'Bayside ', 'Break O Day', 'Brisbane City ', 'Broken Hill City',
       'Broome', 'Bundaberg', 'Bundberg', 'Burdekin', 'Busselton',
       'Byron', 'Cairns', 'Camden', 'Campaspe', 'Canterbury-Bankstown',
       'Carnarvon', 'Cassowary Coast', 'Ceduna', 'Central Desert',
       'Central Highlands', 'Cessnock', 'City of Darwin',
       'City of Griffith', 'City of Kalamunda',
       'City of Kalgoolie-Boulder', 'City of Newcastle', 'City of Perth',
       'City of Wagga Wagga', 'Clarence Valley', 'Cleve', 'Cobar',
       'Colac Otway', 'Cook', 'Dandaragan', 'Derby West Kimberley',
       'Derwent Valley', 'Devonport', 'Douglas', 'Dundas', 'Dungog',
       'East Arnhem', 'East Gippsland', 'East Pilbara ', 'Eurobadalla',
       'Exmouth', 'Flinders', 'Fraser Coast', 'Gladstone',
       'Goldfields - Esperance', 'Greater Bendigo ', 'Greater Shepperon

In [18]:
location_df.query('region == "Bundaberg" or region == "Bundberg"')

site                name      lat      long state     region
66  39059  LADY ELLIOT ISLAND -24.1116  152.7161   QLD  Bundaberg
70  39128      BUNDABERG AERO -24.9069  152.3230   QLD   Bundberg

In [19]:
location_df.query('region == "N.A."')

site                  name      lat      long   state region
32    14703         CENTRE ISLAND -15.7426  136.8192    N.T.   N.A.
45    18115        NEPTUNE ISLAND -35.3365  136.1174    S.A.   N.A.
136  200283         WILLIS ISLAND -16.2878  149.9652  A.E.T.   N.A.
137  200284  COCOS ISLAND AIRPORT -12.1892   96.8344  A.E.T.   N.A.
138  200288   NORFOLK ISLAND AERO -29.0389  167.9408  A.E.T.   N.A.
139  300000                 DAVIS -68.5744   77.9672  A.A.T.   N.A.
140  300004      MACQUARIE ISLAND -54.4994  158.9369  A.A.T.   N.A.

In [20]:
location_df['site'].nunique()

141

##### Observations:
- there are a total of 141 rows in this dataset
- no columns have null values
- we have 141 unique site values as expected
- we have 141 unique site names
- we have 141 unique (lat, long)
- few inconsistent values in the ```state``` column which need to be fixed ('W.A.', 'W.A. ', 'N.T.', 'N.T. ')
- A.A.T. = Australian Antarctic Territory, A.E.T. = Australian External Territories
- in the ```region``` column we have one name ('Bundaberg', 'Bundberg') that has spelling issues. Need to be fixed.

##### Clean-up:

In [21]:
location_df['state'] = location_df['state'].replace({'W.A. ':'W.A.', 'N.T. ':'N.T.'}) # fixing inconsistencies in state col

In [22]:
location_df['state'].unique()

array(['W.A.', 'N.T.', 'S.A.', 'QLD', 'N.S.W.', 'VIC', 'TAS', 'A.E.T.',
       'A.A.T.'], dtype=object)

In [23]:
location_df['region'] = location_df['region'].replace({'Bundberg':'Bundaberg'}) # fixing spelling issue in region col

In [24]:
location_df[location_df['region'].str.startswith('B')]['region']

2                Broome
4                Broome
16            Busselton
38               Barkly
61             Burdekin
66            Bundaberg
68               Banana
70            Bundaberg
72       Brisbane City 
77       Brisbane City 
81     Broken Hill City
87                Byron
94             Bayside 
122            Ballarat
126         Break O Day
127         Break O Day
Name: region, dtype: object

#### description

In [25]:
description_df = pd.read_csv("../datasets/original/description.csv")
description_df

Field                                        Description
0          Location              Bureau of Meteorology Station Number.
1               DMY               Day month year in DD/MM/YYYY format.
2     Precipitation   Precipitation in the 24 hours before 9am (loc...
3        PrecipQual                  * Quality of precipitation value.
4       RainDaysNum   Number of days of rain within the days of acc...
5   RainDaysMeasure   Accumulated number of days over which the pre...
6       Evaporation   Evaporation in 24 hours before 9am (local tim...
7          EvapQual   * Quality of evaporation in 24 hours before 9...
8       EvayDaysNum              Days of accumulation for evaporation.
9           MaxTemp   Maximum temperature in 24 hours after 9am (lo...
10      MaxTempQual   * Quality of maximum temperature in 24 hours ...
11      MaxTempDays       Days of accumulation of maximum temperature.
12          MinTemp   Minimum temperature in 24 hours before 9am (l...
13      MinTempQual   * Quality of minimum temperature in 24 hours ...
14      MinTempDays       Days of accumulation of minimum temperature.
15          Humid00   -*- Relative humidity for observation at 00 h...
16      Humid00Qual   * Quality of relative humidity for observatio...
17          Humid03   -*- Relative humidity for observation at 03 h...
18      Humid03QUal   * Quality of relative humidity for observatio...
19          Humid06   -*- Relative humidity for observation at 06 h...
20      Humid06Qual   * Quality of relative humidity for observatio...
21          Humid09   -*- Relative humidity for observation at 09 h...
22      Humid09Qual   * Quality of relative humidity for observatio...
23          Humid12   -*- Relative humidity for observation at 12 h...
24      Humid12Qual   * Quality of relative humidity for observatio...
25          Humid15   -*- Relative humidity for observation at 15 h...
26      Humid15Qual   * Quality of relative humidity for observatio...
27          Humid18   -*- Relative humidity for observation at 18 h...
28      Humid18Qual   * Quality of relative humidity for observatio...
29          Humid21   -*- Relative humidity for observation at 21 h...
30      Humid21Qual   * Quality of relative humidity for observatio...
31         Sunshine   Number of hours of bright sunshine in the 24 ...
32     SunshineQual                 * Quality of sunshine information.
33           Okta00          Total cloud amount at 00 hours Local Time
34       Okta00Qual   * Quality of total cloud amount at 00 hours L...
35           Okta03          Total cloud amount at 03 hours Local Time
36       Okta03Qual   * Quality of total cloud amount at 03 hours L...
37           Okta06          Total cloud amount at 06 hours Local Time
38       Okta06Qual   * Quality of total cloud amount at 06 hours L...
39           Okta09          Total cloud amount at 09 hours Local Time
40       Okta09Qual   * Quality of total cloud amount at 09 hours L...
41           Okta12          Total cloud amount at 12 hours Local Time
42       Okta12Qual   * Quality of total cloud amount at 12 hours L...
43           Okta15          Total cloud amount at 15 hours Local Time
44       Okta15Qual   * Quality of total cloud amount at 15 hours L...
45           Okta18          Total cloud amount at 18 hours Local Time
46       Okta18Qual   * Quality of total cloud amount at 18 hours L...
47           Okta21          Total cloud amount at 21 hours Local Time
48       Okta21Qual   * Quality of total cloud amount at 21 hours L...

In [26]:
description_df.columns = [col_name.lower() for col_name in description_df] # renaming cols to lower case for conveneince

In [27]:
description_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 49 entries, 0 to 48
Data columns (total 2 columns):
 #   Column       Non-Null Count  Dtype 
---  ------       --------------  ----- 
 0   field        49 non-null     object
 1   description  49 non-null     object
dtypes: object(2)
memory usage: 916.0+ bytes


In [28]:
description_df['field'].unique()

array(['Location', 'DMY', 'Precipitation', 'PrecipQual', 'RainDaysNum',
       'RainDaysMeasure', 'Evaporation', 'EvapQual', 'EvayDaysNum',
       'MaxTemp', 'MaxTempQual', 'MaxTempDays', 'MinTemp', 'MinTempQual',
       'MinTempDays', 'Humid00', 'Humid00Qual', 'Humid03', 'Humid03QUal',
       'Humid06', 'Humid06Qual', 'Humid09', 'Humid09Qual', 'Humid12',
       'Humid12Qual', 'Humid15', 'Humid15Qual', 'Humid18', 'Humid18Qual',
       'Humid21', 'Humid21Qual', 'Sunshine', 'SunshineQual', 'Okta00',
       'Okta00Qual', 'Okta03', 'Okta03Qual', 'Okta06', 'Okta06Qual',
       'Okta09', 'Okta09Qual', 'Okta12', 'Okta12Qual', 'Okta15',
       'Okta15Qual', 'Okta18', 'Okta18Qual', 'Okta21', 'Okta21Qual'],
      dtype=object)

##### Observations:
- total 49 rows in this dataset
- no columns have null values
- Two spelling issues have been noticed in the ```field``` (EvayDaysNum --> EvapDaysNum) and (Humid03QUal --> Humid03Qual)

##### Clean-up:

In [29]:
description_df['field'] = description_df['field'].replace({'EvayDaysNum':'EvapDaysNum','Humid03QUal':'Humid03Qual'})

In [30]:
description_df

field                                        description
0          Location              Bureau of Meteorology Station Number.
1               DMY               Day month year in DD/MM/YYYY format.
2     Precipitation   Precipitation in the 24 hours before 9am (loc...
3        PrecipQual                  * Quality of precipitation value.
4       RainDaysNum   Number of days of rain within the days of acc...
5   RainDaysMeasure   Accumulated number of days over which the pre...
6       Evaporation   Evaporation in 24 hours before 9am (local tim...
7          EvapQual   * Quality of evaporation in 24 hours before 9...
8       EvapDaysNum              Days of accumulation for evaporation.
9           MaxTemp   Maximum temperature in 24 hours after 9am (lo...
10      MaxTempQual   * Quality of maximum temperature in 24 hours ...
11      MaxTempDays       Days of accumulation of maximum temperature.
12          MinTemp   Minimum temperature in 24 hours before 9am (l...
13      MinTempQual   * Quality of minimum temperature in 24 hours ...
14      MinTempDays       Days of accumulation of minimum temperature.
15          Humid00   -*- Relative humidity for observation at 00 h...
16      Humid00Qual   * Quality of relative humidity for observatio...
17          Humid03   -*- Relative humidity for observation at 03 h...
18      Humid03Qual   * Quality of relative humidity for observatio...
19          Humid06   -*- Relative humidity for observation at 06 h...
20      Humid06Qual   * Quality of relative humidity for observatio...
21          Humid09   -*- Relative humidity for observation at 09 h...
22      Humid09Qual   * Quality of relative humidity for observatio...
23          Humid12   -*- Relative humidity for observation at 12 h...
24      Humid12Qual   * Quality of relative humidity for observatio...
25          Humid15   -*- Relative humidity for observation at 15 h...
26      Humid15Qual   * Quality of relative humidity for observatio...
27          Humid18   -*- Relative humidity for observation at 18 h...
28      Humid18Qual   * Quality of relative humidity for observatio...
29          Humid21   -*- Relative humidity for observation at 21 h...
30      Humid21Qual   * Quality of relative humidity for observatio...
31         Sunshine   Number of hours of bright sunshine in the 24 ...
32     SunshineQual                 * Quality of sunshine information.
33           Okta00          Total cloud amount at 00 hours Local Time
34       Okta00Qual   * Quality of total cloud amount at 00 hours L...
35           Okta03          Total cloud amount at 03 hours Local Time
36       Okta03Qual   * Quality of total cloud amount at 03 hours L...
37           Okta06          Total cloud amount at 06 hours Local Time
38       Okta06Qual   * Quality of total cloud amount at 06 hours L...
39           Okta09          Total cloud amount at 09 hours Local Time
40       Okta09Qual   * Quality of total cloud amount at 09 hours L...
41           Okta12          Total cloud amount at 12 hours Local Time
42       Okta12Qual   * Quality of total cloud amount at 12 hours L...
43           Okta15          Total cloud amount at 15 hours Local Time
44       Okta15Qual   * Quality of total cloud amount at 15 hours L...
45           Okta18          Total cloud amount at 18 hours Local Time
46       Okta18Qual   * Quality of total cloud amount at 18 hours L...
47           Okta21          Total cloud amount at 21 hours Local Time
48       Okta21Qual   * Quality of total cloud amount at 21 hours L...

#### state specific BOM data

In [35]:
base_folder = '../datasets/original'
state_specific_BOM_data = ['NT.csv','WA..csv','VIC.csv','AAT.csv','NSW.csv','SA.csv','TAS.csv','AET.csv','QLD.csv']

In [36]:
for file_name in state_specific_BOM_data:
    print(os.path.join(base_folder,file_name))

../datasets/original/NT.csv
../datasets/original/WA..csv
../datasets/original/VIC.csv
../datasets/original/AAT.csv
../datasets/original/NSW.csv
../datasets/original/SA.csv
../datasets/original/TAS.csv
../datasets/original/AET.csv
../datasets/original/QLD.csv


In [37]:
nt_df = pd.read_csv('../datasets/original/NT.csv')
wa_df = pd.read_csv('../datasets/original/WA..csv')
vic_df = pd.read_csv('../datasets/original/VIC.csv')
aat_df = pd.read_csv('../datasets/original/AAT.csv')
nsw_df = pd.read_csv('../datasets/original/NSW.csv')
sa_df = pd.read_csv('../datasets/original/SA.csv')
tas_df = pd.read_csv('../datasets/original/TAS.csv')
aet_df = pd.read_csv('../datasets/original/AET.csv')
qld_df = pd.read_csv('../datasets/original/QLD.csv')

/var/folders/4j/dgknx_s54nn4jt2zpy86mlr00000gn/T/ipykernel_6299/940232725.py:1: DtypeWarning: Columns (2) have mixed types. Specify dtype option on import or set low_memory=False.
  nt_df = pd.read_csv('../datasets/original/NT.csv')
/var/folders/4j/dgknx_s54nn4jt2zpy86mlr00000gn/T/ipykernel_6299/940232725.py:2: DtypeWarning: Columns (9,11,12,14) have mixed types. Specify dtype option on import or set low_memory=False.
  wa_df = pd.read_csv('../datasets/original/WA..csv')


In [38]:
nt_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 242163 entries, 0 to 242162
Data columns (total 49 columns):
 #   Column           Non-Null Count   Dtype 
---  ------           --------------   ----- 
 0   Location         242163 non-null  int64 
 1   DMY              242163 non-null  object
 2   Precipitation    242163 non-null  object
 3   PrecipQual       242163 non-null  object
 4   RainDaysNum      242163 non-null  object
 5   RainDaysMeasure  242163 non-null  object
 6   Evaporation      242163 non-null  object
 7   EvapQual         242163 non-null  object
 8   EvapDaysNum      242163 non-null  object
 9   MaxTemp          242163 non-null  object
 10  MaxTempQual      242163 non-null  object
 11  MaxTempDays      242163 non-null  object
 12  MinTemp          242163 non-null  object
 13  MinTempQual      242163 non-null  object
 14  MinTempDays      242163 non-null  object
 15  Humid00          242163 non-null  object
 16  Humid00Qual      242163 non-null  object
 17  Humid03   

In [39]:
wa_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 447072 entries, 0 to 447071
Data columns (total 49 columns):
 #   Column           Non-Null Count   Dtype 
---  ------           --------------   ----- 
 0   Location         447072 non-null  int64 
 1   DMY              447072 non-null  object
 2   Precipitation    447072 non-null  object
 3   PrecipQual       447072 non-null  object
 4   RainDaysNum      447072 non-null  object
 5   RainDaysMeasure  447072 non-null  object
 6   Evaporation      447072 non-null  object
 7   EvapQual         447072 non-null  object
 8   EvapDaysNum      447072 non-null  object
 9   MaxTemp          447072 non-null  object
 10  MaxTempQual      447072 non-null  object
 11  MaxTempDays      447072 non-null  object
 12  MinTemp          447072 non-null  object
 13  MinTempQual      447072 non-null  object
 14  MinTempDays      447072 non-null  object
 15  Humid00          447072 non-null  object
 16  Humid00Qual      447072 non-null  object
 17  Humid03   

In [40]:
vic_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 380641 entries, 0 to 380640
Data columns (total 49 columns):
 #   Column           Non-Null Count   Dtype 
---  ------           --------------   ----- 
 0   Location         380641 non-null  int64 
 1   DMY              380641 non-null  object
 2   Precipitation    380641 non-null  object
 3   PrecipQual       380641 non-null  object
 4   RainDaysNum      380641 non-null  object
 5   RainDaysMeasure  380641 non-null  object
 6   Evaporation      380641 non-null  object
 7   EvapQual         380641 non-null  object
 8   EvapDaysNum      380641 non-null  object
 9   MaxTemp          380641 non-null  object
 10  MaxTempQual      380641 non-null  object
 11  MaxTempDays      380641 non-null  object
 12  MinTemp          380641 non-null  object
 13  MinTempQual      380641 non-null  object
 14  MinTempDays      380641 non-null  object
 15  Humid00          380641 non-null  object
 16  Humid00Qual      380641 non-null  object
 17  Humid03   

In [41]:
aat_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 37256 entries, 0 to 37255
Data columns (total 49 columns):
 #   Column           Non-Null Count  Dtype 
---  ------           --------------  ----- 
 0   Location         37256 non-null  int64 
 1   DMY              37256 non-null  object
 2   Precipitation    37256 non-null  object
 3   PrecipQual       37256 non-null  object
 4   RainDaysNum      37256 non-null  object
 5   RainDaysMeasure  37256 non-null  object
 6   Evaporation      37256 non-null  object
 7   EvapQual         37256 non-null  object
 8   EvapDaysNum      37256 non-null  object
 9   MaxTemp          37256 non-null  object
 10  MaxTempQual      37256 non-null  object
 11  MaxTempDays      37256 non-null  object
 12  MinTemp          37256 non-null  object
 13  MinTempQual      37256 non-null  object
 14  MinTempDays      37256 non-null  object
 15  Humid00          37256 non-null  object
 16  Humid00Qual      37256 non-null  object
 17  Humid03          37256 non-null

In [42]:
nsw_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 386822 entries, 0 to 386821
Data columns (total 49 columns):
 #   Column           Non-Null Count   Dtype 
---  ------           --------------   ----- 
 0   Location         386822 non-null  int64 
 1   DMY              386822 non-null  object
 2   Precipitation    386822 non-null  object
 3   PrecipQual       386822 non-null  object
 4   RainDaysNum      386822 non-null  object
 5   RainDaysMeasure  386822 non-null  object
 6   Evaporation      386822 non-null  object
 7   EvapQual         386822 non-null  object
 8   EvapDaysNum      386822 non-null  object
 9   MaxTemp          386822 non-null  object
 10  MaxTempQual      386822 non-null  object
 11  MaxTempDays      386822 non-null  object
 12  MinTemp          386822 non-null  object
 13  MinTempQual      386822 non-null  object
 14  MinTempDays      386822 non-null  object
 15  Humid00          386822 non-null  object
 16  Humid00Qual      386822 non-null  object
 17  Humid03   

In [43]:
sa_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 74512 entries, 0 to 74511
Data columns (total 49 columns):
 #   Column           Non-Null Count  Dtype 
---  ------           --------------  ----- 
 0   Location         74512 non-null  int64 
 1   DMY              74512 non-null  object
 2   Precipitation    74512 non-null  object
 3   PrecipQual       74512 non-null  object
 4   RainDaysNum      74512 non-null  object
 5   RainDaysMeasure  74512 non-null  object
 6   Evaporation      74512 non-null  object
 7   EvapQual         74512 non-null  object
 8   EvapDaysNum      74512 non-null  object
 9   MaxTemp          74512 non-null  object
 10  MaxTempQual      74512 non-null  object
 11  MaxTempDays      74512 non-null  object
 12  MinTemp          74512 non-null  object
 13  MinTempQual      74512 non-null  object
 14  MinTempDays      74512 non-null  object
 15  Humid00          74512 non-null  object
 16  Humid00Qual      74512 non-null  object
 17  Humid03          74512 non-null

In [44]:
tas_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 223470 entries, 0 to 223469
Data columns (total 49 columns):
 #   Column           Non-Null Count   Dtype 
---  ------           --------------   ----- 
 0   Location         223470 non-null  int64 
 1   DMY              223470 non-null  object
 2   Precipitation    223470 non-null  object
 3   PrecipQual       223470 non-null  object
 4   RainDaysNum      223470 non-null  object
 5   RainDaysMeasure  223470 non-null  object
 6   Evaporation      223470 non-null  object
 7   EvapQual         223470 non-null  object
 8   EvapDaysNum      223470 non-null  object
 9   MaxTemp          223470 non-null  object
 10  MaxTempQual      223470 non-null  object
 11  MaxTempDays      223470 non-null  object
 12  MinTemp          223470 non-null  object
 13  MinTempQual      223470 non-null  object
 14  MinTempDays      223470 non-null  object
 15  Humid00          223470 non-null  object
 16  Humid00Qual      223470 non-null  object
 17  Humid03   

In [45]:
aet_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 55884 entries, 0 to 55883
Data columns (total 49 columns):
 #   Column           Non-Null Count  Dtype 
---  ------           --------------  ----- 
 0   Location         55884 non-null  int64 
 1   DMY              55884 non-null  object
 2   Precipitation    55884 non-null  object
 3   PrecipQual       55884 non-null  object
 4   RainDaysNum      55884 non-null  object
 5   RainDaysMeasure  55884 non-null  object
 6   Evaporation      55884 non-null  object
 7   EvapQual         55884 non-null  object
 8   EvapDaysNum      55884 non-null  object
 9   MaxTemp          55884 non-null  object
 10  MaxTempQual      55884 non-null  object
 11  MaxTempDays      55884 non-null  object
 12  MinTemp          55884 non-null  object
 13  MinTempQual      55884 non-null  object
 14  MinTempDays      55884 non-null  object
 15  Humid00          55884 non-null  object
 16  Humid00Qual      55884 non-null  object
 17  Humid03          55884 non-null

In [46]:
qld_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 465700 entries, 0 to 465699
Data columns (total 49 columns):
 #   Column           Non-Null Count   Dtype 
---  ------           --------------   ----- 
 0   Location         465700 non-null  int64 
 1   DMY              465700 non-null  object
 2   Precipitation    465700 non-null  object
 3   PrecipQual       465700 non-null  object
 4   RainDaysNum      465700 non-null  object
 5   RainDaysMeasure  465700 non-null  object
 6   Evaporation      465700 non-null  object
 7   EvapQual         465700 non-null  object
 8   EvapDaysNum      465700 non-null  object
 9   MaxTemp          465700 non-null  object
 10  MaxTempQual      465700 non-null  object
 11  MaxTempDays      465700 non-null  object
 12  MinTemp          465700 non-null  object
 13  MinTempQual      465700 non-null  object
 14  MinTempDays      465700 non-null  object
 15  Humid00          465700 non-null  object
 16  Humid00Qual      465700 non-null  object
 17  Humid03   

##### Observations:
- ~~there seems to be no null values in any columns in the above datasets~~
- ~~```DMY``` column need to be changed to datetime column, there are no empty strings~~
- ~~```Humid03QUal``` column name need to be change to ```Humid03Qual``` in all datasets~~
- ~~```RainDaysNum``` to be changed to int column, replace empty strings with pd.NA, find empty strings with .apply(str.isspace)~~
- ~~in ```nt_df```, ```Precipitation``` column as mixed type value (float and str) the rest of the dataframes have values of type str~~
- ~~```PrecipQual``` has empty strings~~
- ~~```RainDaysMeasure``` to be changed to int type, check for empty strings~~
- ~~```'Evaporation'``` to be changed to float type, check for empty strings~~
- ~~```EvapQual``` has empty strings~~
- ~~```EvapDaysNum``` to be changed to int column, replace empty strings with pd.NA~~
- ~~in ```wa_df```, ```MaxTemp``` column as mixed type value (float and str) the rest of the dataframes have values of type str, check for empty strings~~
- ~~```MaxTempQual``` has empty strings~~
- ~~in ```wa_df```, ```'MaxTempDays'``` column as mixed type value (int and str) the rest of the dataframes have values of type str, check for empty strings~~
- ~~in ```wa_df```, ```MinTemp``` column as mixed type value (float and str) the rest of the dataframes have values of type str, check for empty strings~~
- ~~check for empty strings in ```MinTempQual```~~ 
- ~~in ```wa_df```, ```MinTempDays``` column as mixed type value (int and str) the rest of the dataframes have values of type str, check for empty strings~~
- ~~check for empty strings in ```'Humid00'``` change to int type~~
- ~~check for empty strings in ```'Humid00Qual'```~~ 
- ~~check for empty strings in ```'Humid03'``` change to int type`~~
- ~~check for empty strings in ```'Humid03Qual'```~~
- ~~check for empty strings in ```'Humid06'``` change to int type~~
- ~~check for empty strings in ```'Humid06Qual'```~~
- ~~check for empty strings in ```'Humid09'``` change to int type~~
- ~~check for empty strings in ```'Humid09Qual'```~~
- ~~check for empty strings in ```'Humid12'``` change to int type~~
- ~~check for empty strings in ```'Humid12Qual'```~~
- ~~check for empty strings in ```'Humid15'``` change to int type~~
- ~~check for empty strings in ```'Humid15Qual'```~~
- ~~check for empty strings in ```'Humid18'``` change to int type~~
- ~~check for empty strings in ```'Humid18Qual'```~~
- ~~check for empty strings in ```'Humid21'``` change to int type~~
- ~~check for empty strings in ```'Humid21Qual'```~~
- ~~```Sunshine`` col to be changed to float type, check for empty strings~~
- ~~check for empty strings in ```'SunshineQual''```~~
- ~~```'Okta00'``` to be changed to int type, check for empty strings~~
- ~~check for empty strings in ```'Okta00Qual'```~~
- ~~```'Okta03'``` to be changed to int type, check for empty strings~~
- ~~check for empty strings in ```'Okta03Qual'```~~
- ~~```'Okta06'``` to be changed to int type, check for empty strings~~
- ~~check for empty strings in ```'Okta06Qual'```~~
- ~~```'Okta09'``` to be changed to int type, check for empty strings~~
- ~~check for empty strings in ```'Okta09Qual'```~~
- ~~```'Okta12'``` to be changed to int type, check for empty strings~~
- ~~check for empty strings in ```'Okta12Qual'```~~
- ~~```'Okta15'``` to be changed to int type, check for empty strings~~
- ~~check for empty strings in ```'Okta15Qual'```~~
- ~~```'Okta18'``` to be changed to int type, check for empty strings~~
- ~~check for empty strings in ```'Okta18Qual'````~~
- ~~```'Okta21'``` to be changed to int type, check for empty strings~~
- ~~check for empty strings in ```'Okta21Qual'```~~


##### Clean-up:

In [47]:
# cols to be checked for empty strings and changed to int type
to_int = ['RainDaysNum', 'RainDaysMeasure', 'EvapDaysNum',
          'MaxTempDays', 'MinTempDays', 'Humid00',
          'Humid03', 'Humid06', 'Humid09', 'Humid12',
          'Humid15', 'Humid18','Humid21', 'Okta00',
          'Okta03', 'Okta06', 'Okta09', 'Okta12',
          'Okta15','Okta18','Okta21'] 

# cols to be checked for empty strings and changed to float type
to_float = ['Precipitation', 'Evaporation', 
            'MaxTemp', 'MinTemp', 'Sunshine'] 

# cols to be checked for empty strings
other_cols = ['PrecipQual','EvapQual', 'MaxTempQual','MinTempQual',
              'Humid00Qual','Humid03QUal','Humid06Qual', 'Humid09Qual',
              'Humid12Qual', 'Humid15Qual', 'Humid18Qual', 'Humid21Qual',  
              'SunshineQual','Okta00Qual', 'Okta03Qual', 'Okta06Qual',
              'Okta09Qual',  'Okta12Qual',  'Okta15Qual',  'Okta18Qual','Okta21Qual']

In [49]:
# data cleaning helper function
def clean_data(df):
    return (
        df
        # change all columns to str except Location col
        .assign(**{col: df[col].astype(str) for col in df.columns if col != 'Location'}) 
        # now use the str.strip function to remove any unwanted spaces in the data
        .assign(**{col: lambda df_, col=col: df_[col].str.strip() for col in df.columns if col != 'Location'})
        # changes DMY col to datetime type and in YYYY-MM-DD format as sqlite3 understands it in this format
        .assign(
            DMY = lambda df_: pd.to_datetime(df_.DMY, dayfirst=True, errors='coerce') 
        )
        # Change DMY column to string as sqlite3 doesn't have a DATATIME type it treats it as TEXT
        .assign(
            DMY = lambda df_: df_.DMY.astype(str)
        )
        # Change required cols to int while replacing empty strings to NULL value
        .assign(
            **{col: lambda df_, col=col: df_[col].where(df_[col] != '',other=pd.NA).astype('Int64') for col in to_int}
        )
        # Change required cols to float while replacing empty strings to NULL value
        .assign(
            **{col: lambda df_, col=col: df_[col].where(df_[col] != '',other=pd.NA).astype('Float64') for col in to_float}
        )
        # Change required cols to string while replacing empty strings to NULL value
        .assign(
            **{col: lambda df_, col=col: df_[col].where(df_[col] != '',other=pd.NA).astype('string') for col in other_cols}
        )
        .rename(columns={'Humid03QUal':'Humid03Qual'}) # renames mis-spelt col
    )


In [ ]:
# Takes around 1 min to clean all datasets
nt_df= clean_data(nt_df)
wa_df= clean_data(wa_df)
vic_df= clean_data(vic_df)
aat_df= clean_data(aat_df)
nsw_df= clean_data(nsw_df)
sa_df= clean_data(sa_df)
tas_df= clean_data(tas_df)
aet_df= clean_data(aet_df)
qld_df= clean_data(qld_df)

In [51]:
# combining all state wise BOM data
combined_BOM_DATA = pd.concat([nt_df,wa_df,vic_df,aat_df,nsw_df,sa_df,tas_df,aet_df,qld_df],axis=0,ignore_index=True)

In [52]:
combined_BOM_DATA.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2313520 entries, 0 to 2313519
Data columns (total 49 columns):
 #   Column           Dtype  
---  ------           -----  
 0   Location         int64  
 1   DMY              object 
 2   Precipitation    Float64
 3   PrecipQual       string 
 4   RainDaysNum      Int64  
 5   RainDaysMeasure  Int64  
 6   Evaporation      Float64
 7   EvapQual         string 
 8   EvapDaysNum      Int64  
 9   MaxTemp          Float64
 10  MaxTempQual      string 
 11  MaxTempDays      Int64  
 12  MinTemp          Float64
 13  MinTempQual      string 
 14  MinTempDays      Int64  
 15  Humid00          Int64  
 16  Humid00Qual      string 
 17  Humid03          Int64  
 18  Humid03Qual      string 
 19  Humid06          Int64  
 20  Humid06Qual      string 
 21  Humid09          Int64  
 22  Humid09Qual      string 
 23  Humid12          Int64  
 24  Humid12Qual      string 
 25  Humid15          Int64  
 26  Humid15Qual      string 
 27  Humid18     

##### Exporting ```bom_data``` table

In [53]:
# paraquet allows us to save space when dealing with larger file sizes
combined_BOM_DATA.to_parquet('../datasets/final_datasets/BOM_DATA_FINAL.parquet', compression='gzip', index=False)

### Creation/ Modification of Dataset(s)

#### Exploration

In [54]:
location_df

site                  name      lat      long   state  \
0      1006          WYNDHAM AERO -15.5100  128.1503    W.A.   
1      1007      TROUGHTON ISLAND -13.7542  126.1485    W.A.   
2      3003        BROOME AIRPORT -17.9475  122.2352    W.A.   
3      3032            DERBY AERO -17.3706  123.6611    W.A.   
4      4019               MANDORA -19.7419  120.8433    W.A.   
..      ...                   ...      ...       ...     ...   
136  200283         WILLIS ISLAND -16.2878  149.9652  A.E.T.   
137  200284  COCOS ISLAND AIRPORT -12.1892   96.8344  A.E.T.   
138  200288   NORFOLK ISLAND AERO -29.0389  167.9408  A.E.T.   
139  300000                 DAVIS -68.5744   77.9672  A.A.T.   
140  300004      MACQUARIE ISLAND -54.4994  158.9369  A.A.T.   

                     region  
0    Wyndham East Kimberley  
1                 Kimberley  
2                    Broome  
3      Derby West Kimberley  
4                    Broome  
..                      ...  
136                    N.A.  
137                    N.A.  
138                    N.A.  
139                    N.A.  
140                    N.A.  

[141 rows x 6 columns]

In [55]:
unique_regions = location_df['region'].unique() # checking if a region appears in multiple states
for region in unique_regions:
    if location_df.query(f'region == "{region}"')['state'].nunique() > 1:
        print(region)

N.A.


In [56]:
location_df.query(' region == "N.A."')['state'].unique()

array(['N.T.', 'S.A.', 'A.E.T.', 'A.A.T.'], dtype=object)

- only N.A. most likely Not Available appears in multiple states, needs to be handled individually

#### ```states``` table

In [57]:
unique_states = location_df['state'].unique()
unique_states.sort()
unique_states

array(['A.A.T.', 'A.E.T.', 'N.S.W.', 'N.T.', 'QLD', 'S.A.', 'TAS', 'VIC',
       'W.A.'], dtype=object)

In [58]:
states_df = pd.DataFrame({
    'state_id': np.arange(len(unique_states)) + 1,
    'state': unique_states })

states_df = states_df.assign(
    state_id = lambda df_: df_['state_id'].astype(int),
    state = lambda df_: df_['state'].astype('string'),
).rename(columns={'state':'state_name'})
states_df

state_id state_name
0         1     A.A.T.
1         2     A.E.T.
2         3     N.S.W.
3         4       N.T.
4         5        QLD
5         6       S.A.
6         7        TAS
7         8        VIC
8         9       W.A.

#### ```regions``` table

In [59]:
temp_regions_df = location_df[['region','state']].drop_duplicates()
temp_regions_df = temp_regions_df.sort_values(by='region', ascending=True)

regions_df = temp_regions_df.assign(
    region_id = lambda df_: np.arange(df_['region'].size) + 1
).rename(columns={'state':'state_id', 'region': 'region_name'})[['region_id','region_name','state_id']]

regions_df

region_id                    region_name state_id
55           1  Aboriginal Shire of Kowanyama      QLD
49           2                       Adelaide     S.A.
39           3                  Alice Springs     N.T.
15           4       Augusta - Margaret River     W.A.
122          5                       Ballarat      VIC
..         ...                            ...      ...
113        115                     Wellington      VIC
51         116                   West Torrens     S.A.
29         117                 Western Arnhem     N.T.
47         118                        Whyalla     S.A.
0          119         Wyndham East Kimberley     W.A.

[119 rows x 3 columns]

In [60]:
regions_df.query(' region_name == "N.A." ')

region_id region_name state_id
139         85        N.A.   A.A.T.
136         86        N.A.   A.E.T.
32          87        N.A.     N.T.
45          88        N.A.     S.A.

In [61]:
regions_df = regions_df.assign(
    state_id = regions_df['state_id'].replace(dict(states_df[['state_name','state_id']].values))
)
regions_df

/var/folders/4j/dgknx_s54nn4jt2zpy86mlr00000gn/T/ipykernel_6299/641880813.py:2: FutureWarning: Downcasting behavior in `replace` is deprecated and will be removed in a future version. To retain the old behavior, explicitly call `result.infer_objects(copy=False)`. To opt-in to the future behavior, set `pd.set_option('future.no_silent_downcasting', True)`
  state_id = regions_df['state_id'].replace(dict(states_df[['state_name','state_id']].values))


region_id                    region_name  state_id
55           1  Aboriginal Shire of Kowanyama         5
49           2                       Adelaide         6
39           3                  Alice Springs         4
15           4       Augusta - Margaret River         9
122          5                       Ballarat         8
..         ...                            ...       ...
113        115                     Wellington         8
51         116                   West Torrens         6
29         117                 Western Arnhem         4
47         118                        Whyalla         6
0          119         Wyndham East Kimberley         9

[119 rows x 3 columns]

#### ```weather_stations``` table

In [62]:
location_df = location_df.assign(
    state = lambda df_: df_['state'].replace(dict(states_df[['state_name','state_id']].values)),
    region = lambda df_: df_['region'].replace(dict(regions_df[['region_name','region_id']].values))
).rename(columns={'site':'site_id', 'state':'state_id', 'region':'region_id'})
location_df

/var/folders/4j/dgknx_s54nn4jt2zpy86mlr00000gn/T/ipykernel_6299/3827063515.py:2: FutureWarning: Downcasting behavior in `replace` is deprecated and will be removed in a future version. To retain the old behavior, explicitly call `result.infer_objects(copy=False)`. To opt-in to the future behavior, set `pd.set_option('future.no_silent_downcasting', True)`
  state = lambda df_: df_['state'].replace(dict(states_df[['state_name','state_id']].values)),
/var/folders/4j/dgknx_s54nn4jt2zpy86mlr00000gn/T/ipykernel_6299/3827063515.py:3: FutureWarning: Downcasting behavior in `replace` is deprecated and will be removed in a future version. To retain the old behavior, explicitly call `result.infer_objects(copy=False)`. To opt-in to the future behavior, set `pd.set_option('future.no_silent_downcasting', True)`
  region = lambda df_: df_['region'].replace(dict(regions_df[['region_name','region_id']].values))


site_id                  name      lat      long  state_id  region_id
0       1006          WYNDHAM AERO -15.5100  128.1503         9        119
1       1007      TROUGHTON ISLAND -13.7542  126.1485         9         69
2       3003        BROOME AIRPORT -17.9475  122.2352         9         12
3       3032            DERBY AERO -17.3706  123.6611         9         40
4       4019               MANDORA -19.7419  120.8433         9         12
..       ...                   ...      ...       ...       ...        ...
136   200283         WILLIS ISLAND -16.2878  149.9652         2         88
137   200284  COCOS ISLAND AIRPORT -12.1892   96.8344         2         88
138   200288   NORFOLK ISLAND AERO -29.0389  167.9408         2         88
139   300000                 DAVIS -68.5744   77.9672         1         88
140   300004      MACQUARIE ISLAND -54.4994  158.9369         1         88

[141 rows x 6 columns]

#### ```bom_data_metadata``` table

In [63]:
description_df.head()

field                                        description
0       Location              Bureau of Meteorology Station Number.
1            DMY               Day month year in DD/MM/YYYY format.
2  Precipitation   Precipitation in the 24 hours before 9am (loc...
3     PrecipQual                  * Quality of precipitation value.
4    RainDaysNum   Number of days of rain within the days of acc...

- this will be a stan alone table for reference

#### ```quality_metadata``` table

In [64]:
with open('../datasets/original/quality-metatdata.txt', 'r') as fp:
    qf_data = fp.readlines()

In [65]:
qf_data[3:]

['Y: quality controlled and acceptable\n',
 'N: not quality controlled\n',
 'W: quality controlled and considered wrong\n',
 'S: quality controlled and considered suspect\n',
 'I: quality controlled and inconsistent with other known information\n',
 'X: no quality information available\n']

In [66]:
quality_flags_data = []
for entry in qf_data[3:]:
    quality_flags_data.append(entry.split(':'))
quality_flags_data

[['Y', ' quality controlled and acceptable\n'],
 ['N', ' not quality controlled\n'],
 ['W', ' quality controlled and considered wrong\n'],
 ['S', ' quality controlled and considered suspect\n'],
 ['I', ' quality controlled and inconsistent with other known information\n'],
 ['X', ' no quality information available\n']]

In [67]:
quality_flags_df = pd.DataFrame(quality_flags_data, columns=['quality_flag','description'])
quality_flags_df

quality_flag                                        description
0            Y                quality controlled and acceptable\n
1            N                           not quality controlled\n
2            W          quality controlled and considered wrong\n
3            S        quality controlled and considered suspect\n
4            I   quality controlled and inconsistent with othe...
5            X                 no quality information available\n

In [68]:
quality_flags_df = quality_flags_df.assign(
    description = lambda df_: df_['description'].str.strip()
)
quality_flags_df

quality_flag                                        description
0            Y                  quality controlled and acceptable
1            N                             not quality controlled
2            W            quality controlled and considered wrong
3            S          quality controlled and considered suspect
4            I  quality controlled and inconsistent with other...
5            X                   no quality information available

#### Exporting data

In [69]:
states_df.to_parquet('../datasets/final_datasets/states.parquet', compression='gzip', index=False)

In [70]:
regions_df.to_parquet('../datasets/final_datasets/regions.parquet', compression='gzip', index=False)


In [71]:
location_df.to_parquet('../datasets/final_datasets/weather_stations.parquet', compression='gzip', index=False)

In [72]:
description_df.to_parquet('../datasets/final_datasets/bom_data_metadata.parquet', compression='gzip', index=False)

In [73]:
quality_flags_df.to_parquet('../datasets/final_datasets/quality_metadata.parquet', compression='gzip', index=False)

### Verify Final Datasets

In [74]:
list_dir('../datasets/final_datasets/')

../datasets/final_datasets/ contain the following folders: NO FOLDERS FOUND IN DIRECTORY
../datasets/final_datasets/ contain the following files: ['regions.parquet', 'weather_stations.parquet', 'BOM_DATA_FINAL.parquet', 'bom_data_metadata.parquet', 'quality_metadata.parquet', 'states.parquet']


In [75]:
pd.read_parquet('../datasets/final_datasets/regions.parquet').head()


region_id                    region_name  state_id
0          1  Aboriginal Shire of Kowanyama         5
1          2                       Adelaide         6
2          3                  Alice Springs         4
3          4       Augusta - Margaret River         9
4          5                       Ballarat         8

In [76]:
pd.read_parquet('../datasets/final_datasets/weather_stations.parquet').head()

site_id              name      lat      long  state_id  region_id
0     1006      WYNDHAM AERO -15.5100  128.1503         9        119
1     1007  TROUGHTON ISLAND -13.7542  126.1485         9         69
2     3003    BROOME AIRPORT -17.9475  122.2352         9         12
3     3032        DERBY AERO -17.3706  123.6611         9         40
4     4019           MANDORA -19.7419  120.8433         9         12

In [77]:
pd.read_parquet('../datasets/final_datasets/BOM_DATA_FINAL.parquet').head()

Location         DMY  Precipitation PrecipQual  RainDaysNum  \
0     14015  1970-01-01            0.0          Y         <NA>   
1     14015  1970-01-02            0.0          Y         <NA>   
2     14015  1970-01-03            0.0          Y         <NA>   
3     14015  1970-01-04            0.5          Y            1   
4     14015  1970-01-05            0.0          Y         <NA>   

   RainDaysMeasure  Evaporation EvapQual  EvapDaysNum  MaxTemp  ... Okta09  \
0             <NA>         10.2        Y            1     34.3  ...      5   
1             <NA>         10.9        Y            1     33.4  ...      8   
2             <NA>          8.4        Y            1     31.1  ...      8   
3                1          4.3        Y            1     33.3  ...      8   
4             <NA>          7.4        Y            1     34.6  ...      6   

   Okta09Qual  Okta12 Okta12Qual  Okta15  Okta15Qual Okta18  Okta18Qual  \
0           Y       4          Y       4           Y      8           Y   
1           Y       8          Y       8           Y      8           Y   
2           Y       8          Y       8           Y      8           Y   
3           Y       6          Y       5           Y      7           Y   
4           Y       7          Y       6           Y      4           Y   

  Okta21  Okta21Qual  
0      8           Y  
1      8           Y  
2      6           Y  
3      5           Y  
4      1           Y  

[5 rows x 49 columns]

In [78]:
pd.read_parquet('../datasets/final_datasets/bom_data_metadata.parquet').head()

field                                        description
0       Location              Bureau of Meteorology Station Number.
1            DMY               Day month year in DD/MM/YYYY format.
2  Precipitation   Precipitation in the 24 hours before 9am (loc...
3     PrecipQual                  * Quality of precipitation value.
4    RainDaysNum   Number of days of rain within the days of acc...

In [79]:
pd.read_parquet('../datasets/final_datasets/quality_metadata.parquet').head()

quality_flag                                        description
0            Y                  quality controlled and acceptable
1            N                             not quality controlled
2            W            quality controlled and considered wrong
3            S          quality controlled and considered suspect
4            I  quality controlled and inconsistent with other...

In [80]:
pd.read_parquet('../datasets/final_datasets/states.parquet').head()

state_id state_name
0         1     A.A.T.
1         2     A.E.T.
2         3     N.S.W.
3         4       N.T.
4         5        QLD